### PR TITLE
Add precommit and run on all files

### PR DESCRIPTION
**Pull request by thomasyu888**
Adds precommit so that development goes easier

### CODE REVIEW

File: .github/workflows/test.yml
Comment (GiaJordan) [inline .github/workflows/test.yml:1]: Should we also enforce `black` for all files now with this change?

File: .pre-commit-config.yaml
Comment (GiaJordan) [inline .pre-commit-config.yaml:21]: Are the comments hooks we plan to turn on in the future? are we no longer using `isort` as well?

File: .github/workflows/test.yml
Comment (thomasyu888) [reply .github/workflows/test.yml:1]: The `.precommit.config..` has this (buried under the sea of file changes) ha.

```
  - repo: https://github.com/psf/black
    rev: 23.12.1
    hooks:
      - id: black
        language_version: python3
```

So it does enforce black with this change.

File: .pre-commit-config.yaml
Comment (thomasyu888) [reply .pre-commit-config.yaml:21]: ^ so isort seems to cause a weird error with the code, so I turned it off for now. Will turn it back on later.

Comment (GiaJordan) [reply .pre-commit-config.yaml:21]: that makes sense, sounds good

File: .github/workflows/test.yml
Comment (GiaJordan) [reply .github/workflows/test.yml:1]: Okay cool, so this will now run the precommit hook in the action itself and modify the files instead of just checking that the files are already formatted correctly?

Comment (thomasyu888) [reply .github/workflows/test.yml:1]: So once this is merged in there will be a new development workflow.  Everyone will run precommit install locally and whenever you commit, you'll something like this:

```
git commit -am "Use pre-commit"
[INFO] Installing environment for https://github.com/pre-commit/pre-commit-hooks.
[INFO] Once installed this environment will be reused.
[INFO] This may take a few minutes...
trim trailing whitespace.................................................Passed
check for added large files..............................................Passed
check python ast.....................................(no files to check)Skipped
check for merge conflicts................................................Passed
check xml............................................(no files to check)Skipped
check yaml...............................................................Passed
debug statements (python)............................(no files to check)Skipped
fix end of files.........................................................Passed
fix requirements.txt.................................(no files to check)Skipped
mixed line ending........................................................Passed
black................................................(no files to check)Skipped
[add-precommit d1741be0] Use pre-commit
 1 file changed, 1 insertion(+), 5 deletions(-)
```

It will automatically run these for you during commit.  This pre-commit hook is just an added layer and will fail if pre-commit isn't installed locally.

**Review by BryanFauble — APPROVED** [review]
Hopefully isort is an easy fix, I like not needing to think about adding imports in certain locations - No biggie though.


Thanks for running this on all files, current PR's are going to run into merge conflicts because of the sheer number of changes, but, it's a bandage we need to rip off at some point.

**Comment by thomasyu888** [comment]
> Hopefully isort is an easy fix, I like not needing to think about adding imports in certain locations - No biggie though.
> 
> Thanks for running this on all files, current PR's are going to run into merge conflicts because of the sheer number of changes, but, it's a bandage we need to rip off at some point.

I'll wait for all your PRs to go in, and I'm happy to fix the merge conflicts since it'll be easier to fix on my end

File: .pre-commit-config.yaml
Comment (thomasyu888) [reply .pre-commit-config.yaml:21]: Ya, we can certainly look into it!  My concern is always that `ruff` may clash against `black` or other formatting tools, but thats a concern i have about every formatting tool. 😓

**Review by GiaJordan — APPROVED** [review]

**Comment by BWMac** [comment]
#1560 is merged, good to go when ready.

**Review by BryanFauble — APPROVED** [review]

**Pull request merged**
